src/grain/harvest.js

```diff
@@ -0,0 +1,220 @@
+    throw new Error(`invalid harvestAmount: ${String(harvestAmount)}`);
+  }
+
+  const totalCred = sum(Array.from(cred.values()));
+  if (totalCred === 0) {
+    return [];
+  }
+
+  return mapToArray(cred, ([address, cred]) => ({
+    address,
+    amount: multiplyFloat(harvestAmount, cred / totalCred),
+  }));
+}
+
+/**
+ * Distribute a fixed amount of Grain to the users who were "most underpaid".
+ *
+ * We consider a user underpaid if they have recieved a smaller proportion of
+ * past earnings than their share of score. They are fairly paid if their
+ * proportion of earnings is equal to their score share, and they are overpaid
+ * if their proportion of earnings is higher than their share of the score.
+ *
+ * We start by imagining a hypothetical world, where the entire grain supply of
+ * the project (including this harvest) were distributed according to the
+ * current scores. Based on this, we can calculate the "fair" lifetime earnings
+ * for each participant. Usually, some will be "underpaid" (they recieved less
+ * than this amount) and others are "overpaid".
+ *
+ * We can sum across all users who were underpaid to find the "total
+ * underpayment". As an invariant, `totalUnderpayment = harvestAmount +
+ * totalOverpayment`, since no-one has yet been paid the harvestAmount, and
+ * beyond that every grain of overpayment to one actor is underpayment for a
+ * different actor.
+ *
+ * Now that we've calculated each actor's underpayment, and the total
+ * underpayment, we divide the harvest's grain amount across users in
+ * proportion to their underpayment.
+ *
+ * You should use this harvest when you want to divide a fixed amount of grain
+ * across participants in a way that aligns long-term payment with total cred
+ * scores.
+ */
+function computeFairReceipts(
+  harvestAmount: Grain,
+  credMap: Map<NodeAddressT, number>,
+  earnings: Map<NodeAddressT, Grain>
+): $ReadOnlyArray<GrainReceipt> {
+  let totalEarnings = ZERO;
+  for (const e of earnings.values()) {
+    totalEarnings += e;
+  }
+  let totalCred = 0;
+  for (const s of credMap.values()) {
+    totalCred += s;
+  }
+  if (totalCred === 0) {
+    return [];
+  }
+
+  const targetGrainPerCred = multiplyFloat(
+    totalEarnings + harvestAmount,
+    1 / totalCred
+  );
+
+  let totalUnderpayment = ZERO;
+  const userUnderpayment: Map<NodeAddressT, Grain> = new Map();
+  const addresses = new Set([...credMap.keys(), ...earnings.keys()]);
+
+  for (const addr of addresses) {
+    const earned = earnings.get(addr) || ZERO;
+    const cred = credMap.get(addr) || 0;
+
+    const target = multiplyFloat(targetGrainPerCred, cred);
+    if (target > earned) {
+      const underpayment = target - earned;
+      userUnderpayment.set(addr, underpayment);
+      totalUnderpayment += underpayment;
+    }
+  }
+
+  return mapToArray(userUnderpayment, ([address, underpayment]) => {
+    const underpaymentProportion =
+      Number(underpayment) / Number(totalUnderpayment);
```

Comment:
We should add a math helper method for this so that we can use our (naive) implementation and @wchargin can come and write something more sophisticated later.

I'm fine adding the helper method in this PR, I don't think we need to split it into another.